src/tools/dym/superinstructions/CandidateDetector.java

```diff
@@ -0,0 +1,164 @@
+    // [C_0, s_0, C_1, s_1, C_2]
+    // We now find possible "piblings", i.e. siblings of its parent.
+    // For that, we invoke findExtensions on the prefix
+    // [C_0]
+    Map<Integer, ActivationContext> piblings = findExtensions(
+        new Object[] {
+            currentContext.getClass(0)
+        });
+    // Also, we find possible siblings. For that, we invoke findExtensions
+    // on the prefix
+    // [C_0, s_0, C_1]
+    Map<Integer, ActivationContext> siblings = findExtensions(
+        new Object[] {
+            currentContext.getClass(0),
+            currentContext.getChildIndex(0),
+            currentContext.getClass(1)
+        });
+    // We now construct a super-instruction candidate, i.e.
+    // a tree of height 2. The root of the tree is C_0 (its Java type is unknown).
+    Candidate candidate = new Candidate(currentContext.getClass(0), "?");
+    // Now, we add the children of C_0, i.e. the siblings of C_1 and C_1 itself.
+    for (int piblingSlot : piblings.keySet()) {
+      if (piblingSlot == currentContext.getChildIndex(0)) {
+        // This is C_1. We add it to the candidate and proceed
+        // with adding the siblings of C_2 and C_2 itself.
+        Candidate.Node child =
+            candidate.getRoot().setChild(piblingSlot, currentContext.getClass(1), "?");
+        for (int siblingSlot : siblings.keySet()) {
+          if (siblingSlot == currentContext.getChildIndex(1)) {
+            // Add C_2
+            child.setChild(siblingSlot,
+                currentContext.getClass(2),
+                currentContext.getJavaType());
+          } else {
+            // Add a sibling of C_2
+            ActivationContext sibling = siblings.get(siblingSlot);
+            child.setChild(siblingSlot,
+                sibling.getClass(2),
+                sibling.getJavaType());
+          }
+        }
+      } else {
+        ActivationContext pibling = piblings.get(piblingSlot);
+        // Add a sibling of C_1.
+        assert pibling.getNumberOfClasses() == 2;
+        candidate.getRoot().setChild(piblingSlot,
+            pibling.getClass(1),
+            pibling.getJavaType());
+      }
+    }
+    // The score of the super-instruction candidate corresponds to the number
+    // of activations of the current context.
+    candidate.setScore(contexts.get(currentContext));
+    return candidate;
+  }
+
+  /**
+   * Run the candidate detection heuristic and return a report as a String.
+   */
+  public String detect() {
+    // Find all activation contexts with traces of length n = 2,
+    // exclude traces containing SequenceNodes (they do not yield
+    // viable candidates for super-instructions),
+    // sort them by activation counts in descending order,
+    // fetch the top CONSIDER_TOP_CONTEXTS activation contexts
+    // and construct a candidate for each of them.
+    List<ActivationContext> sorted =
+        contexts.keySet().stream().filter(context -> context.getNumberOfClasses() == 3)
```

Comment:
@fredreichbier here I was wondering why that's a three. The comment above talks about 2, I believe.

Reply:
Oops you're right, there is an inconsistency here.
Given a trace `[C_0, i_0, ... C_n]`, `getNumberOfClasses` does not actually return `n`, it returns `n+1` (because e.g. a trace of length 1 is `[C_0, i_0, C_1]`, so `n=1`, but `getNumberOfClasses() == 2`). I guess the most sensible way to resolve this would be to modify the documentation of `getNumberOfClasses`.